src/components/HistoryRouter.tsx

```diff
@@ -0,0 +1,48 @@
+import React, { useLayoutEffect } from 'react';
+import { HistoryRouterProps, Router } from 'react-router-dom';
+import { Update } from 'history';
+
+/** Strips leading "!" from paths */
+const normalizePath = (pathname: string) => pathname.replace(/^!/, '');
+
+/**
+ * A slightly customized version of the HistoryRouter from react-router-dom.
+ * We need to use HistoryRouter to have a shared history state between react-router and appRouter, but it does not seem
+ * to be properly exported in the upstream package.
+ * We also needed some customizations to handle #! routes.
+ * Refs: https://github.com/remix-run/react-router/blob/v6.3.0/packages/react-router-dom/index.tsx#L222
+ */
+export function HistoryRouter({ basename, children, history }: HistoryRouterProps) {
+    const [state, setState] = React.useState({
```

Comment:
```suggestion
    const [state, setState] = React.useState<Update>({
```
_Lint should warn on type, imo, but it doesn't_ :man_shrugging: